builtins: implement pg_table_is_visible as UDF

jordanlewis opened the pull request:
Now that we support UDFs, we can implement builtin functions as "virtual
UDFs", that are defined in the builtins map.

The builtin appears to be about twice as fast with this method, which is
nice because it gets used a lot in ORM introspection queries.

Release note (performance improvement): improve the performance of
pg_table_is_visible

Epic: None

Co-authored-by: rafiss <rafi@cockroachlabs.com>

jordanlewis committed:
builtins: implement pg_*_is_visible as UDF

Now that we support UDFs, we can implement builtin functions as "virtual
UDFs", that are defined in the builtins map.

The builtin appears to be about twice as fast with this method, which is
nice because it gets used a lot in ORM introspection queries.

Epic: None

Release note (performance improvement): improve the performance of
pg_{function,table,type}_is_visible

Co-authored-by: rafiss <rafi@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94339)
<!-- Reviewable:end -->

jordanlewis commented:
UDF team - is this a valid thing to do? Do you understand why the performance would be improved? @mgartner @chengxiong-ruan

mgartner reviewed (approved):
Seems valid to me. Perhaps inefficiencies in the internal executor explain the speed up from this change?

rafiss reviewed (changes requested):
want to get `pg_type_is_visible` and `pg_function_is_visible` while you're here?

jordanlewis committed:
builtins: implement _pg_index_position as UDF

Release note: None

jordanlewis commented:
Done

jordanlewis committed:
sql: hide x-db objs in type,func virtual indexes

Previously, the pg_proc and pg_type virtual indexes would show cross-db
objects, which is not supposed to happen. This is now corrected.

Release note (bug fix): the pg_proc and pg_type virtual oid indexes no
longer incorrectly show cross-db objects. This is unlikely to affect
real-world use cases.

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[`pkg/sql/pg_catalog.go` line 3335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94339#-NKUJTOiBs0lIgumxPxN:-NKUJTOiBs0lIgumxPxO:bgck49w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2456880a236bad0a22782e816bd0804af3ce4b06/pkg/sql/pg_catalog.go#L3335)):*
> ```Go
> 				// table. We have special logic for this case.
> 				if typDesc.GetKind() == descpb.TypeDescriptor_TABLE_IMPLICIT_RECORD_TYPE {
> 					table, err := p.Descriptors().GetImmutableTableByID(ctx, p.txn, id, tree.ObjectLookupFlags{})
> ```

does this part also need to check for dropped tables?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94339)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/pg_catalog.go` line 3335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94339#-NKUJTOiBs0lIgumxPxN:-NKUN4RP3py3oKIkkmoC:bfcq7zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2456880a236bad0a22782e816bd0804af3ce4b06/pkg/sql/pg_catalog.go#L3335)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does this part also need to check for dropped tables?
</blockquote></details>

Good point, fixed


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94339#-:-NKURRGQAn8HJOutkSSn:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/catalog/typedesc/table_implicit_record_type.go:
```diff
@@ -175,7 +175,6 @@ func (v TableImplicitRecordType) Adding() bool {
 // Dropped implements the Descriptor interface.
 func (v TableImplicitRecordType) Dropped() bool {
-	v.panicNotSupported("Dropped")
 	return false
```
Shouldn't this instead delegate to the wrapped table descriptor?

jordanlewis replied:
Thanks, done

mgartner reviewed (approved):
LGTM

jordanlewis commented:
bors r+
